File: mobile/modules/bluetooth-sdk/android/src/main/java/com/mentra/bluetoothsdk/BluetoothSdk.kt
package com.mentra.core

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.mentra.core.utils.PhoneAudioMonitor
import java.util.Collections

class BluetoothSdk private constructor(
    context: Context,
    private val config: BluetoothSdkConfig,
    listener: BluetoothSdkListener,
) : AutoCloseable {
    private val appContext = context.applicationContext
    private val mainHandler = Handler(Looper.getMainLooper())
    private val deviceManager: CoreManager
    private val listeners =
        Collections.synchronizedSet(mutableSetOf<BluetoothSdkListener>())
    private val discoveredDeviceNames = mutableSetOf<String>()
    private val bridgeEventSinkId: String
    private val storeListenerId: String

    init {
        listeners.add(listener)
        Bridge.initialize(appContext)
        deviceManager = CoreManager.getInstance()
        bridgeEventSinkId = Bridge.addEventSink { eventName, data -> dispatchBridgeEvent(eventName, data) }
        storeListenerId = GlassesStore.store.addListener { category, changes -> dispatchStoreUpdate(category, changes) }
    }

    companion object {
        @JvmStatic
        fun create(
            context: Context,
            listener: BluetoothSdkListener,
        ): BluetoothSdk = create(context, BluetoothSdkConfig(), listener)

        @JvmStatic
        fun create(
            context: Context,
            config: BluetoothSdkConfig,
            listener: BluetoothSdkListener,
        ): BluetoothSdk = BluetoothSdk(context, config, listener)
    }

    fun addListener(listener: BluetoothSdkListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: BluetoothSdkListener) {
        listeners.remove(listener)
    }

    fun getGlassesStatus(): MentraGlassesStatus =
        MentraGlassesStatus(GlassesStore.store.getCategory("glasses"))

    fun getBluetoothStatus(): MentraBluetoothStatus =
        MentraBluetoothStatus(GlassesStore.store.getCategory(ObservableStore.CORE_CATEGORY))

    fun startScan(model: MentraDeviceModel) {
        discoveredDeviceNames.clear()
        GlassesStore.apply(ObservableStore.CORE_CATEGORY, "searching", true)
        deviceManager.findCompatibleDevices(model.deviceType)
    }

    fun stopScan() {
        GlassesStore.apply(ObservableStore.CORE_CATEGORY, "searching", false)
        dispatchToListeners { it.onScanStopped(MentraScanStopReason.CANCELLED) }
    }

    fun connect(device: MentraDiscoveredDevice) {
        GlassesStore.apply(ObservableStore.CORE_CATEGORY, "pending_wearable", device.model.deviceType)
        deviceManager.connectByName(device.name)
    }

    fun connectByName(model: MentraDeviceModel, deviceName: String) {
        GlassesStore.apply(ObservableStore.CORE_CATEGORY, "pending_wearable", model.deviceType)
        deviceManager.connectByName(deviceName)
    }

    fun connectByName(deviceName: String) {
        deviceManager.connectByName(deviceName)
    }

    fun connectDefault() {
        deviceManager.connectDefault()
    }

    fun connectSimulated() {
        deviceManager.connectSimulated()
    }

    fun disconnect() {
        deviceManager.disconnect()
    }

    fun forget() {
        deviceManager.forget()
    }

    fun displayText(request: MentraDisplayTextRequest) {
        deviceManager.displayText(request.toMap())
    }

    fun displayEvent(request: MentraDisplayEventRequest) {
        deviceManager.displayEvent(request.toMap())
    }

    fun clearDisplay() {
        deviceManager.clearDisplay()
    }

    fun showDashboard() {
        deviceManager.showDashboard()
    }
    
    fun setOwnAppAudioPlaying(playing: Boolean) {
        PhoneAudioMonitor.getInstance(appContext).setOwnAppAudioPlaying(playing)
    }

    fun requestWifiScan() {
        deviceManager.requestWifiScan()
    }

    fun sendWifiCredentials(ssid: String, password: String) {
        deviceManager.sendWifiCredentials(ssid, password)
    }

    fun forgetWifiNetwork(ssid: String) {
        deviceManager.forgetWifiNetwork(ssid)
    }

    fun setHotspotState(enabled: Boolean) {
        deviceManager.setHotspotState(enabled)
    }

    fun requestPhoto(request: MentraPhotoRequest) {
        deviceManager.photoRequest(
            request.requestId,
            request.appId,
            request.size,
            request.webhookUrl,
            request.authToken,
            request.compress,
            request.flash,
            request.sound,
        )
    }

    fun queryGalleryStatus() {
        deviceManager.queryGalleryStatus()
    }

    fun startStream(request: MentraStreamRequest) {
        deviceManager.startStream(request.values.toMutableMap())
    }

    fun keepStreamAlive(request: MentraStreamKeepAliveRequest) {
        deviceManager.keepStreamAlive(request.values.toMutableMap())
    }

    fun stopStream() {
        deviceManager.stopStream()
    }

    fun startVideoRecording(request: MentraVideoRecordingRequest) {
        deviceManager.startVideoRecording(request.requestId, request.save, request.flash, request.sound)
    }

    fun stopVideoRecording(requestId: String) {
        deviceManager.stopVideoRecording(requestId)
    }

    fun requestVersionInfo() {
        deviceManager.requestVersionInfo()
    }

    fun sendOtaStart() {
        deviceManager.sendOtaStart()
    }

    fun sendShutdown() {
        deviceManager.sendShutdown()
    }

    fun sendReboot() {
        deviceManager.sendReboot()
    }

    fun sendIncidentId(incidentId: String, apiBaseUrl: String? = null) {
        deviceManager.sendIncidentId(incidentId, apiBaseUrl)
    }

    override fun close() {
        Bridge.removeEventSink(bridgeEventSinkId)
        GlassesStore.store.removeListener(storeListenerId)
        listeners.clear()
    }

    private fun dispatchStoreUpdate(category: String, changes: Map<String, Any>) {
        when (ObservableStore.normalizeCategory(category)) {
            "glasses" ->
                dispatchToListeners {
                    it.onGlassesStatusChanged(MentraGlassesStatusUpdate(changes))
                }
            ObservableStore.CORE_CATEGORY -> {
                dispatchToListeners {
                    it.onBluetoothStatusChanged(MentraBluetoothStatusUpdate(changes))
                }
                dispatchDiscoveredDevices(changes["searchResults"])
            }
        }
    }

    private fun dispatchDiscoveredDevices(rawSearchResults: Any?) {
        val results = rawSearchResults as? List<*> ?: return
        results.forEach { rawResult ->
            val result = rawResult as? Map<*, *> ?: return@forEach
            val name = result["deviceName"] as? String ?: result["name"] as? String ?: return@forEach
            if (!discoveredDeviceNames.add(name)) return@forEach
            val model = MentraDeviceModel.fromDeviceType(result["deviceModel"] as? String)
            val device = MentraDiscoveredDevice(model = model, name = name)
            dispatchToListeners { it.onDeviceDiscovered(device) }
        }
    }

    private fun dispatchBridgeEvent(eventName: String, data: Map<String, Any>) {
        when (eventName) {
            "log" -> dispatchToListeners { it.onLog(data["message"] as? String ?: data.toString()) }
            "button_press" ->
                dispatchToListeners {
                    it.onButtonPress(
                        MentraButtonPressEvent(
                            buttonId = data["buttonId"] as? String ?: "",
                            pressType = data["pressType"] as? String ?: "",
                            timestamp = (data["timestamp"] as? Number)?.toLong(),
                        )
                    )
                }
            "touch_event" -> dispatchToListeners { it.onTouch(MentraTouchEvent(data)) }
            "head_up" -> dispatchToListeners { it.onHeadUpChanged(data["up"] as? Boolean ?: false) }
            "battery_status" ->
                dispatchToListeners {
                    it.onBatteryStatus(
                        MentraBatteryStatusEvent(
                            level = (data["level"] as? Number)?.toInt(),
                            charging = data["charging"] as? Boolean,
                            values = data,
                        )
                    )
                }
            "wifi_status_change" -> dispatchToListeners { it.onWifiStatusChanged(MentraWifiStatusEvent(data)) }
            "gallery_status" -> dispatchToListeners { it.onGalleryStatus(MentraGalleryStatusEvent(data)) }
            "photo_response" -> dispatchToListeners { it.onPhotoResponse(MentraPhotoResponseEvent(data)) }
            "stream_status" -> dispatchToListeners { it.onStreamStatus(MentraStreamStatusEvent(data)) }
            "mic_pcm" -> (data["pcm"] as? ByteArray)?.let { frame ->
                dispatchToListeners { it.onMicPcm(frame) }
            }
            "mic_lc3" -> (data["lc3"] as? ByteArray)?.let { frame ->
                dispatchToListeners { it.onMicLc3(frame) }
            }
            "local_transcription" ->
                dispatchToListeners {
                    it.onLocalTranscription(
                        MentraLocalTranscriptionEvent(
                            text = data["text"] as? String ?: "",
                            isFinal = data["isFinal"] as? Boolean ?: false,
                            values = data,
                        )
                    )
                }
            "compatible_glasses_search_stop" ->
                dispatchToListeners { it.onScanStopped(MentraScanStopReason.COMPLETED) }
            "pair_failure" ->
                dispatchToListeners {
                    it.onError(
                        MentraBluetoothError(
                            code = "pair_failure",
                            message = data["error"] as? String ?: data.toString(),
                        )
                    )
                }
            else -> dispatchToListeners { it.onRawEvent(eventName, data) }
        }
    }

    private fun dispatchToListeners(callback: (BluetoothSdkListener) -> Unit) {
        val snapshot = synchronized(listeners) { listeners.toList() }
        val deliver = {
            snapshot.forEach { listener ->
                try {
                    callback(listener)
                } catch (error: Throwable) {
                    // Listener exceptions should not crash Bluetooth event delivery.
                }
            }
        }
        if (config.deliverCallbacksOnMainThread && Looper.myLooper() != Looper.getMainLooper()) {
            mainHandler.post(deliver)
        } else {
            deliver()
        }
    }
}

File: mobile/modules/bluetooth-sdk/ios/Source/BluetoothSDK.swift
import Foundation

public struct BluetoothSDKConfiguration {
    public static let `default` = BluetoothSDKConfiguration()

    public init() {}
}

public enum DeviceModel: String {
    case g1
    case g2
    case Live
    case Nex
    case mach1
    case z100
    case frame
    case simulated
    case r1

    var deviceType: String {
        switch self {
        case .g1:
            DeviceTypes.G1
        case .g2:
            DeviceTypes.G2
        case .Live:
            DeviceTypes.LIVE
        case .Nex:
            DeviceTypes.NEX
        case .mach1:
            DeviceTypes.MACH1
        case .z100:
            DeviceTypes.Z100
        case .frame:
            DeviceTypes.FRAME
        case .simulated:
            DeviceTypes.SIMULATED
        case .r1:
            ControllerTypes.R1
        }
    }

    static func fromDeviceType(_ deviceType: String?) -> DeviceModel {
        switch deviceType {
        case DeviceTypes.G1:
            .g1
        case DeviceTypes.G2:
            .g2
        case DeviceTypes.LIVE:
            .Live
        case DeviceTypes.NEX:
            .Nex
        case DeviceTypes.MACH1:
            .mach1
        case DeviceTypes.Z100:
            .z100
        case DeviceTypes.FRAME:
            .frame
        case DeviceTypes.SIMULATED:
            .simulated
        case ControllerTypes.R1:
            .r1
        default:
            .Live
        }
    }
}

public struct DiscoveredDevice: CustomStringConvertible {
    public let model: DeviceModel
    public let name: String
    public let identifier: String?
    public let rssi: Int?

    public init(
        model: DeviceModel,
        name: String,
        identifier: String? = nil,
        rssi: Int? = nil
    ) {
        self.model = model
        self.name = name
        self.identifier = identifier
        self.rssi = rssi
    }

    public var description: String {
        "DiscoveredDevice(model: \(model), name: \(name))"
    }
}

public struct PairedDevice: CustomStringConvertible {
    public let model: DeviceModel
    public let name: String
    public let identifier: String?

    public init(model: DeviceModel, name: String, identifier: String? = nil) {
        self.model = model
        self.name = name
        self.identifier = identifier
    }

    public var description: String {
        "PairedDevice(model: \(model), name: \(name))"
    }
}

public struct GlassesStatus: CustomStringConvertible {
    public let values: [String: Any]

    public init(values: [String: Any]) {
        self.values = values
    }

    public var description: String {
        values.description
    }
}

public struct BluetoothStatus: CustomStringConvertible {
    public let values: [String: Any]

    public init(values: [String: Any]) {
        self.values = values
    }

    public var description: String {
        values.description
    }
}

public struct GlassesStatusUpdate: CustomStringConvertible {
    public let values: [String: Any]

    public init(values: [String: Any]) {
        self.values = values
    }

    public var description: String {
        values.description
    }
}

public struct BluetoothStatusUpdate: CustomStringConvertible {
    public let values: [String: Any]

    public init(values: [String: Any]) {
        self.values = values
    }

    public var description: String {
        values.description
    }
}

public struct DisplayTextRequest {
    public let text: String
    public let x: Int
    public let y: Int
    public let size: Int

    public init(text: String, x: Int = 0, y: Int = 0, size: Int = 24) {
        self.text = text
        self.x = x
        self.y = y
        self.size = size
    }

    var dictionary: [String: Any] {
        [
            "text": text,
            "x": x,
            "y": y,
            "size": size,
        ]
    }
}

public struct DisplayEventRequest {
    public let values: [String: Any]

    public init(values: [String: Any]) {
        self.values = values
    }
}

public struct DashboardPositionRequest {
    public let height: Int
    public let depth: Int

    public init(height: Int, depth: Int) {
        self.height = height
        self.depth = depth
    }
}

public struct DashboardMenuItem {
    public let title: String
    public let packageName: String
    public let values: [String: Any]

    public init(title: String, packageName: String, values: [String: Any] = [:]) {
        self.title = title
        self.packageName = packageName
        self.values = values
    }

    var dictionary: [String: Any] {
        values.merging(["title": title, "packageName": packageName]) { _, new in new }
    }
}

public enum GalleryMode {
    case auto
    case manual
}

public enum ButtonMode: String {
    case photo
    case video
    case none
}

public enum PhotoSize: String {
    case small
    case medium
    case large
}

public struct ButtonPhotoSettings {
    public let size: PhotoSize

    public init(size: PhotoSize) {
        self.size = size
    }
}

public struct ButtonVideoRecordingSettings {
    public let width: Int
    public let height: Int
    public let fps: Int

    public init(width: Int, height: Int, fps: Int) {
        self.width = width
        self.height = height
        self.fps = fps
    }
}

public enum CameraFov {
    case standard
    case wide

    var value: [String: Int] {
        switch self {
        case .standard:
            ["fov": 118, "roi_position": 0]
        case .wide:
            ["fov": 118, "roi_position": 0]
        }
    }
}

public struct MicConfiguration {
    public let sendPcmData: Bool
    public let sendTranscript: Bool
    public let bypassVad: Bool
    public let sendLc3Data: Bool

    public init(
        sendPcmData: Bool,
        sendTranscript: Bool,
        bypassVad: Bool,
        sendLc3Data: Bool = false
    ) {
        self.sendPcmData = sendPcmData
        self.sendTranscript = sendTranscript
        self.bypassVad = bypassVad
        self.sendLc3Data = sendLc3Data
    }
}

public enum MicPreference: String {
    case auto
    case phone
    case glasses
    case bluetooth
}

public struct PhotoRequest {
    public let requestId: String
    public let appId: String
    public let size: String
    public let webhookUrl: String?
    public let authToken: String?
    public let compress: String?
    public let flash: Bool
    public let sound: Bool

    public init(
        requestId: String,
        appId: String,
        size: String,
        webhookUrl: String? = nil,
        authToken: String? = nil,
        compress: String? = nil,
        flash: Bool,
        sound: Bool
    ) {
        self.requestId = requestId
        self.appId = appId
        self.size = size
        self.webhookUrl = webhookUrl
        self.authToken = authToken
        self.compress = compress
        self.flash = flash
        self.sound = sound
    }
}

public struct StreamRequest {
    public let values: [String: Any]

    public init(values: [String: Any]) {
        self.values = values
    }
}

public struct StreamKeepAliveRequest {
    public let values: [String: Any]

    public init(values: [String: Any]) {
        self.values = values
    }
}

public struct VideoRecordingRequest {
    public let requestId: String
    public let save: Bool
    public let flash: Bool
    public let sound: Bool

    public init(requestId: String, save: Bool, flash: Bool, sound: Bool) {
        self.requestId = requestId
        self.save = save
        self.flash = flash
        self.sound = sound
    }
}

public struct ButtonPressEvent: CustomStringConvertible {
    public let buttonId: String
    public let pressType: String
    public let timestamp: Int?

    public init(buttonId: String, pressType: String, timestamp: Int? = nil) {
        self.buttonId = buttonId
        self.pressType = pressType
        self.timestamp = timestamp
    }

    public var description: String {
        "ButtonPressEvent(buttonId: \(buttonId), pressType: \(pressType))"
    }
}

public struct BluetoothError: Error, CustomStringConvertible {
    public let code: String
    public let message: String

    public init(code: String, message: String) {
        self.code = code
        self.message = message
    }

    public var description: String {
        "\(code): \(message)"
    }
}

public enum ScanStopReason {
    case completed
    case cancelled
    case error
}

public struct LocalTranscriptionEvent: CustomStringConvertible {
    public let text: String
    public let isFinal: Bool
    public let values: [String: Any]

    public init(text: String, isFinal: Bool, values: [String: Any]) {
        self.text = text
        self.isFinal = isFinal
        self.values = values
    }

    public var description: String {
        "LocalTranscriptionEvent(text: \(text), isFinal: \(isFinal))"
    }
}

public enum BluetoothEvent: CustomStringConvertible {
    case localTranscription(LocalTranscriptionEvent)
    case raw(name: String, values: [String: Any])

    public var description: String {
        switch self {
        case let .localTranscription(event):
            event.description
        case let .raw(name, values):
            "\(name): \(values)"
        }
    }
}

@MainActor
public protocol BluetoothSDKDelegate: AnyObject {
    func BluetoothSDK(_ sdk: BluetoothSDK, didUpdateGlassesStatus status: GlassesStatusUpdate)
    func BluetoothSDK(_ sdk: BluetoothSDK, didUpdateBluetoothStatus status: BluetoothStatusUpdate)
    func BluetoothSDK(_ sdk: BluetoothSDK, didDiscover device: DiscoveredDevice)
    func BluetoothSDK(_ sdk: BluetoothSDK, didStopScan reason: ScanStopReason)
    func BluetoothSDK(_ sdk: BluetoothSDK, didReceive event: BluetoothEvent)
    func BluetoothSDK(_ sdk: BluetoothSDK, didReceiveMicPcm frame: Data)
    func BluetoothSDK(_ sdk: BluetoothSDK, didReceiveMicLc3 frame: Data)
    func BluetoothSDK(_ sdk: BluetoothSDK, didChangeDefaultDevice device: PairedDevice?)
    func BluetoothSDK(_ sdk: BluetoothSDK, didLog message: String)
    func BluetoothSDK(_ sdk: BluetoothSDK, didFail error: BluetoothError)
}

@MainActor
public extension BluetoothSDKDelegate {
    func BluetoothSDK(_: BluetoothSDK, didUpdateGlassesStatus _: GlassesStatusUpdate) {}
    func BluetoothSDK(_: BluetoothSDK, didUpdateBluetoothStatus _: BluetoothStatusUpdate) {}
    func BluetoothSDK(_: BluetoothSDK, didDiscover _: DiscoveredDevice) {}
    func BluetoothSDK(_: BluetoothSDK, didStopScan _: ScanStopReason) {}
    func BluetoothSDK(_: BluetoothSDK, didReceive _: BluetoothEvent) {}
    func BluetoothSDK(_: BluetoothSDK, didReceiveMicPcm _: Data) {}
    func BluetoothSDK(_: BluetoothSDK, didReceiveMicLc3 _: Data) {}
    func BluetoothSDK(_: BluetoothSDK, didChangeDefaultDevice _: PairedDevice?) {}
    func BluetoothSDK(_: BluetoothSDK, didLog _: String) {}
    func BluetoothSDK(_: BluetoothSDK, didFail _: BluetoothError) {}
}

@MainActor
public final class BluetoothSDK {
    public weak var delegate: BluetoothSDKDelegate?

    private let configuration: BluetoothSDKConfiguration
    private var discoveredDeviceNames = Set<String>()
    private var bridgeEventSinkId: String?
    private var storeListenerId: String?

    public init(configuration: BluetoothSDKConfiguration = .default) {
        self.configuration = configuration
        bridgeEventSinkId = Bridge.addEventSink { [weak self] eventName, data in
            Task { @MainActor [weak self] in
                self?.dispatchBridgeEvent(eventName, data)
            }
        }
        storeListenerId = GlassesStore.shared.store.addListener { [weak self] category, changes in
            self?.dispatchStoreUpdate(category, changes)
        }
    }

    public var glassesStatus: GlassesStatus {
        GlassesStatus(values: GlassesStore.shared.store.getCategory("glasses"))
    }

    public var bluetoothStatus: BluetoothStatus {
        BluetoothStatus(values: GlassesStore.shared.store.getCategory(ObservableStore.coreCategory))
    }

    public func startScan(model: DeviceModel) {
        discoveredDeviceNames.removeAll()
        GlassesStore.shared.apply(ObservableStore.coreCategory, "searching", true)
        CoreManager.shared.findCompatibleDevices(model.deviceType)
    }

    public func stopScan() {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "searching", false)
        delegate?.BluetoothSDK(self, didStopScan: .cancelled)
    }

    public func connect(to device: DiscoveredDevice) {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "pending_wearable", device.model.deviceType)
        CoreManager.shared.connectByName(device.name)
    }

    public func connect(model: DeviceModel, name: String) {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "pending_wearable", model.deviceType)
        CoreManager.shared.connectByName(name)
    }

    public func connectByName(_ name: String) {
        CoreManager.shared.connectByName(name)
    }

    public func connectDefault() {
        CoreManager.shared.connectDefault()
    }

    public func connectSimulated() {
        CoreManager.shared.connectSimulated()
    }

    public func disconnect() {
        CoreManager.shared.disconnect()
    }

    public func forget() {
        CoreManager.shared.forget()
    }

    public func displayText(_ request: DisplayTextRequest) async throws {
        CoreManager.shared.displayText(request.dictionary)
    }

    public func displayEvent(_ request: DisplayEventRequest) async throws {
        CoreManager.shared.displayEvent(request.values)
    }

    public func clearDisplay() async throws {
        CoreManager.shared.sgc?.clearDisplay()
    }

    public func showDashboard() {
        CoreManager.shared.showDashboard()
    }

    public func setBrightness(_ level: Int, autoMode: Bool? = nil) async throws {
        if let autoMode {
            GlassesStore.shared.apply(ObservableStore.coreCategory, "auto_brightness", autoMode)
        }
        GlassesStore.shared.apply(ObservableStore.coreCategory, "brightness", level)
    }

    public func setAutoBrightness(enabled: Bool) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "auto_brightness", enabled)
    }

    public func setDashboardPosition(_ request: DashboardPositionRequest) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "dashboard_height", request.height)
        GlassesStore.shared.apply(ObservableStore.coreCategory, "dashboard_depth", request.depth)
    }

    public func setDashboardMenu(_ items: [DashboardMenuItem]) async throws {
        GlassesStore.shared.apply(
            ObservableStore.coreCategory,
            "dashboard_menu_apps",
            items.map(\.dictionary)
        )
    }

    public func setHeadUpAngle(_ angleDegrees: Int) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "head_up_angle", angleDegrees)
    }

    public func setScreenDisabled(_ disabled: Bool) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "screen_disabled", disabled)
    }

    public func setGalleryMode(_ mode: GalleryMode) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "gallery_mode", mode == .auto)
    }

    public func setButtonMode(_ mode: ButtonMode) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "button_mode", mode.rawValue)
    }

    public func setButtonPhotoSettings(_ settings: ButtonPhotoSettings) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "button_photo_size", settings.size.rawValue)
    }

    public func setButtonVideoRecordingSettings(_ settings: ButtonVideoRecordingSettings) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "button_video_width", settings.width)
        GlassesStore.shared.apply(ObservableStore.coreCategory, "button_video_height", settings.height)
        GlassesStore.shared.apply(ObservableStore.coreCategory, "button_video_fps", settings.fps)
    }

    public func setButtonCameraLed(enabled: Bool) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "button_camera_led", enabled)
    }

    public func setButtonMaxRecordingTime(minutes: Int) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "button_max_recording_time", minutes)
    }

    public func setCameraFov(_ fov: CameraFov) async throws {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "camera_fov", fov.value)
    }

    public func setMicState(_ config: MicConfiguration) {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "should_send_pcm", config.sendPcmData)
        GlassesStore.shared.apply(ObservableStore.coreCategory, "should_send_lc3", config.sendLc3Data)
        GlassesStore.shared.apply(ObservableStore.coreCategory, "should_send_transcript", config.sendTranscript)
        GlassesStore.shared.apply(ObservableStore.coreCategory, "bypass_vad", config.bypassVad)
        CoreManager.shared.setMicState()
    }

    public func setPreferredMic(_ preferredMic: MicPreference) {
        GlassesStore.shared.apply(ObservableStore.coreCategory, "preferred_mic", preferredMic.rawValue)
    }

    public func setOwnAppAudioPlaying(_ playing: Bool) {
        PhoneAudioMonitor.getInstance().setOwnAppAudioPlaying(playing)
    }

    public func requestWifiScan() {
        CoreManager.shared.requestWifiScan()
    }

    public func sendWifiCredentials(ssid: String, password: String) {
        CoreManager.shared.sendWifiCredentials(ssid, password)
    }

    public func forgetWifiNetwork(ssid: String) {
        CoreManager.shared.forgetWifiNetwork(ssid)
    }

    public func setHotspotState(enabled: Bool) {
        CoreManager.shared.setHotspotState(enabled)
    }

    public func requestPhoto(_ request: PhotoRequest) {
        CoreManager.shared.photoRequest(
            request.requestId,
            request.appId,
            request.size,
            request.webhookUrl,
            request.authToken,
            request.compress,
            request.flash,
            request.sound
        )
    }

    public func queryGalleryStatus() {
        CoreManager.shared.queryGalleryStatus()
    }

    public func startStream(_ request: StreamRequest) {
        CoreManager.shared.startStream(request.values)
    }

    public func keepStreamAlive(_ request: StreamKeepAliveRequest) {
        CoreManager.shared.keepStreamAlive(request.values)
    }

    public func stopStream() {
        CoreManager.shared.stopStream()
    }

    public func startVideoRecording(_ request: VideoRecordingRequest) {
        CoreManager.shared.startVideoRecording(
            request.requestId,
            request.save,
            request.flash,
            request.sound
        )
    }

    public func stopVideoRecording(requestId: String) {
        CoreManager.shared.stopVideoRecording(requestId)
    }

    public func requestVersionInfo() {
        CoreManager.shared.requestVersionInfo()
    }

    public func sendOtaStart() {
        CoreManager.shared.sendOtaStart()
    }

    public func sendShutdown() {
        CoreManager.shared.sendShutdown()
    }

    public func sendReboot() {
        CoreManager.shared.sendReboot()
    }

    public func sendIncidentId(_ incidentId: String, apiBaseUrl: String? = nil) {
        CoreManager.shared.sendIncidentId(incidentId, apiBaseUrl: apiBaseUrl)
    }

    public func invalidate() {
        if let bridgeEventSinkId {
            Bridge.removeEventSink(bridgeEventSinkId)
            self.bridgeEventSinkId = nil
        }
        if let storeListenerId {
            GlassesStore.shared.store.removeListener(storeListenerId)
            self.storeListenerId = nil
        }
        delegate = nil
    }

    private func dispatchStoreUpdate(_ category: String, _ changes: [String: Any]) {
        switch ObservableStore.normalizeCategory(category) {
        case "glasses":
            delegate?.BluetoothSDK(self, didUpdateGlassesStatus: GlassesStatusUpdate(values: changes))
        case ObservableStore.coreCategory:
            delegate?.BluetoothSDK(self, didUpdateBluetoothStatus: BluetoothStatusUpdate(values: changes))
            dispatchDiscoveredDevices(changes["searchResults"])
        default:
            break
        }
    }

    private func dispatchDiscoveredDevices(_ rawSearchResults: Any?) {
        guard let results = rawSearchResults as? [[String: Any]] else { return }
        for result in results {
            guard let name = result["deviceName"] as? String ?? result["name"] as? String else { continue }
            guard discoveredDeviceNames.insert(name).inserted else { continue }
            let model = DeviceModel.fromDeviceType(result["deviceModel"] as? String)
            let device = DiscoveredDevice(model: model, name: name)
            delegate?.BluetoothSDK(self, didDiscover: device)
        }
    }

    private func dispatchBridgeEvent(_ eventName: String, _ data: [String: Any]) {
        switch eventName {
        case "log":
            delegate?.BluetoothSDK(self, didLog: data["message"] as? String ?? data.description)
        case "mic_pcm":
            if let frame = data["pcm"] as? Data {
                delegate?.BluetoothSDK(self, didReceiveMicPcm: frame)
            }
        case "mic_lc3":
            if let frame = data["lc3"] as? Data {
                delegate?.BluetoothSDK(self, didReceiveMicLc3: frame)
            }
        case "local_transcription":
            let event = LocalTranscriptionEvent(
                text: data["text"] as? String ?? "",
                isFinal: data["isFinal"] as? Bool ?? false,
                values: data
            )
            delegate?.BluetoothSDK(self, didReceive: .localTranscription(event))
        case "compatible_glasses_search_stop":
            delegate?.BluetoothSDK(self, didStopScan: .completed)
        case "pair_failure":
            delegate?.BluetoothSDK(
                self,
                didFail: BluetoothError(
                    code: "pair_failure",
                    message: data["error"] as? String ?? data.description
                )
            )
        default:
            delegate?.BluetoothSDK(self, didReceive: .raw(name: eventName, values: data))
        }
    }
}